relay-server/src/services/buffer/envelope_store/sqlite.rs

```diff
@@ -24,35 +25,78 @@ use sqlx::{Pool, QueryBuilder, Row, Sqlite};
 use tokio::fs::DirBuilder;
 use tokio::time::sleep;
 
+const ZSTD_MAGIC_WORD: &[u8] = &[40, 181, 47, 253];
+
 /// Struct that contains all the fields of an [`Envelope`] that are mapped to the database columns.
+#[derive(Clone, Debug)]
 pub struct InsertEnvelope {
     received_at: i64,
     own_key: ProjectKey,
     sampling_key: ProjectKey,
     encoded_envelope: Vec<u8>,
 }
 
+#[derive(Debug, thiserror::Error)]
+pub enum InsertEnvelopeError {
+    #[error("envelope conversion error: {0}")]
+    Envelope(#[from] EnvelopeError),
+    #[error("compression error: {0}")]
+    Zstd(#[from] std::io::Error),
+}
+
+impl InsertEnvelope {
+    // TODO: explain why 1
+    const COMPRESSION_LEVEL: i32 = 1;
+
+    pub fn len(&self) -> usize {
+        self.encoded_envelope.len()
+    }
+
+    pub fn received_at(&self) -> Instant {
+        StartTime::from_timestamp_millis(self.received_at as u64).into_inner()
+    }
+}
+
+impl TryFrom<InsertEnvelope> for Box<Envelope> {
+    type Error = InsertEnvelopeError;
+
+    fn try_from(value: InsertEnvelope) -> Result<Self, Self::Error> {
+        let InsertEnvelope {
+            received_at,
+            own_key,
+            sampling_key,
+            mut encoded_envelope,
+        } = value;
+
+        if encoded_envelope.starts_with(ZSTD_MAGIC_WORD) {
+            encoded_envelope = zstd::decode_all(encoded_envelope.as_slice())?;
```

Comment:
Can we keep a metric to measure decoding speed? Or we are confident from our previous results that it is good enough?